crates/wasi/src/preview2/pipe.rs

```diff
@@ -7,227 +7,835 @@
-    pub fn new(w: W) -> Self {
-        Self::from_shared(Arc::new(RwLock::new(w)))
+impl AsyncWriteStream {
+    /// Create a [`AsyncWriteStream`]. In order to use the [`HostOutputStream`] impl
+    /// provided by this struct, the argument must impl [`tokio::io::AsyncWrite`].
+    pub fn new<T: tokio::io::AsyncWrite + Send + Sync + Unpin + 'static>(mut writer: T) -> Self {
+        let (sender, mut receiver) = tokio::sync::mpsc::channel::<Bytes>(1);
+        let (result_sender, result_receiver) = tokio::sync::mpsc::channel(1);
+
+        crate::preview2::spawn(async move {
+            'outer: loop {
+                use tokio::io::AsyncWriteExt;
+                match receiver.recv().await {
+                    Some(mut bytes) => {
+                        while !bytes.is_empty() {
+                            match writer.write_buf(&mut bytes).await {
+                                Ok(0) => {
+                                    let _ = result_sender.send(Ok(StreamState::Closed)).await;
+                                    break 'outer;
+                                }
+                                Ok(_) => {
+                                    if bytes.is_empty() {
+                                        match result_sender.send(Ok(StreamState::Open)).await {
+                                            Ok(_) => break,
+                                            Err(_) => break 'outer,
+                                        }
+                                    }
+                                    continue;
+                                }
+                                Err(e) => {
+                                    let _ = result_sender.send(Err(e)).await;
+                                    break 'outer;
+                                }
+                            }
+                        }
+                    }
+
+                    // The other side of the channel hung up, the task can exit now
+                    None => break 'outer,
+                }
+            }
+        });
+
+        AsyncWriteStream {
+            state: Some(WriteState::Ready),
+            sender,
+            result_receiver,
+        }
     }
 
-    /// Create a new pipe from a shareable `Write` type.
-    ///
-    /// All `Handle` write operations delegate to writing to this underlying writer.
-    pub fn from_shared(writer: Arc<RwLock<W>>) -> Self {
-        Self { writer }
+    fn send(&mut self, bytes: Bytes) -> anyhow::Result<(usize, StreamState)> {
+        use tokio::sync::mpsc::error::TrySendError;
+
+        debug_assert!(matches!(self.state, Some(WriteState::Ready)));
+
+        let len = bytes.len();
+        match self.sender.try_send(bytes) {
+            Ok(_) => {
+                self.state = Some(WriteState::Pending);
+                Ok((len, StreamState::Open))
+            }
+            Err(TrySendError::Full(_)) => Ok((0, StreamState::Open)),
```

Comment:
Should this case be `unreachable!` since if the state is "ready" then the task should always be trying to receive one buffer?

Reply:
correct- that was left over from a prior factoring of this code, but its unreachable now